registry/modules/authBasic/controllers/authBasic.ts

```diff
@@ -18,109 +18,92 @@ import {
+  login: (props: BasicAuthSchema) => Promise<AuthOutput>;
+  signup: (props: BasicAuthSchema) => Promise<AuthOutput>;
+  refreshToken: (props: RefreshTokenSchema) => Promise<TokensOutput>;
 }
 
 export const createAuthBasicController = (
   userRepo: UserRepository,
   refreshTokenRepo: RefreshTokenRepository
 ): AuthBasicController => {
-  const generateAccessToken = (userId: number) => {
-    const signedJWT = accessTokenManager.sign({ userId: userId.toString() });
+  const generateTokens = async (userId: number, tokenFamily?: string) => {
+    const [refreshToken, accessToken] = await Promise.all([
+      generateRefreshToken(userId, tokenFamily),
+      generateAccessToken(userId),
+    ]);
+    return { accessToken, refreshToken };
+  };
 
-    return signedJWT;
+  const generateAccessToken = (userId: number) => {
+    return accessTokenManager.sign({ userId: userId.toString() });
   };
 
   const generateRefreshToken = async (userId: number, tokenFamily?: string) => {
-    const expAt = new Date(new Date().getTime() + 31 * 24 * 60 * 60000); // Expire in 31 days
-    const refreshTokenExp = expAt.toISOString();
-
-    const token = await refreshTokenRepo.createToken({
+    const expAt = new Date(Date.now() + 31 * 24 * 60 * 60 * 1000);
+    return refreshTokenRepo.createToken({
       userId,
       tokenFamily,
-      expiresAt: refreshTokenExp,
+      expiresAt: expAt.toISOString(),
     });
+  };
 
-    return token;
+  const getUserOrFail = async (username: string) => {
+    const user = await userRepo.getUser(username);
+    if (!user) throw invalidLoginCredentials();
+    return user;
   };
 
   return {
-    async signup(props) {
-      const { username, password, email } = props;
-
-      await userRepo.getUser(username).then((res) => {
-        if (res !== null) throw usernameNotAvailable();
-      });
-
-      // timeCost, parallelism and memoryCost configured according to OWASP recommendations: https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html
-      const hashedPass = await argon.hash(password, {
-        timeCost: 2,
-        parallelism: 1,
-        memoryCost: 19456, // 19 MiB
-      });
-
-      const newUser = await userRepo.createAuthBasicUser({
-        username,
-        hashedPass,
-        email,
-      });
-
-      const refreshToken = await generateRefreshToken(newUser.userId);
-      const accessToken = generateAccessToken(newUser.userId);
+    async signup({ username, password, email }) {
+      if (await userRepo.getUser(username)) throw usernameNotAvailable();
+
+      const [hashedPass, newUser] = await Promise.all([
+        argon2.hash(password, {
+          type: argon2.argon2id,
+          timeCost: 2,
+          parallelism: 1,
+          memoryCost: 12288, // 12 MiB for better performance/security balance
+        }),
+        userRepo.createAuthBasicUser({ username, email }),
+      ]);
+
+      newUser.password = hashedPass;
```

Comment:
This won't work as `userRepo.createAuthBasicUser({ username, email })` expects a `hashedPass` property. Mutating it after creation won't sore it in the DB :). Please keep it as it was for now as we can't parallelise this part!